### ISSUE TITLE
Heavy allocations in UnboundLambdaState.CanonicallyCompareDiagnostics

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#33846]
**Version Used**: 15.9.7

:link: Originally filed as https://developercommunity.visualstudio.com/content/problem/463943/significant-wait-dialogs-all-projects.html

`UnboundLambdaState.CanonicallyCompareDiagnostics` is responsible for nearly 8GB of allocations (14% of total) in this trace. Nearly all of this number comes from `CSDiagnostic.ToString()` calling `DiagnosticFormatter.Format`. Nested within those we find:

* `CSharp.ErrorFacts.GetMessage` calls `Enum.ToString` (4.5GB)
* `DiagnosticFormatter.Format` calls `string.Format` (1.6GB)
* `DiagnosticInfo.GetMessage` calls `string.Format` (0.64GB)
* `DiagnosticFormatter.Format` calls `DiagnosticFormatter.FormatSourceSpan` (0.4GB)
* `DiagnosticFormatter.Format` calls `DiagnosticFormatter.GetMessagePrefix` (0.4GB)

**Comment by CyrusNajmabadi** [comment]
AH, good old enum.tostring.  I thought the runtime had finally done something about that guy.

**Comment by YairHalberstadt** [comment]
The ironic thing is the message is completely irrelevant here. All we care about is picking some arbitrary, consistent way of ordering diagnostics so that the tests for roslyn don't constantly have to be rewritten.

I think the thing to do is just remove the call to format, and explicitly check for enough of the things that make up the message so that all the tests currently work without making any changes.

**Comment by YairHalberstadt** [comment]
// The question then rears its head: when given two of "the same" error
            // to report that are nevertheless different in their arguments, which one
            // do we choose? To the user it hardly matters; either one points to the
            // right location in source code. But it surely matters to our testing team;
            // we do not want to be in a position where some small change to our internal
            // representation of lambdas causes tests to break because errors are reported
            // differently.
            //
            // What we need to do is find a *repeatable* arbitrary way to choose between
            // two errors; we can for example simply take the one that is lower in alphabetical
            // order when converted to a string.

**Comment by YairHalberstadt** [comment]
Converting from the current:
```csharp
           return codeCompare == 0 ? string.CompareOrdinal(x.ToString(), y.ToString()) : codeCompare;
```
to
```csharp
            if (codeCompare != 0)
                return codeCompare;

            for (int i = 0; i < x.Arguments.Count && i < y.Arguments.Count; i++)
            {
                object argx = x.Arguments[i];
                object argy = y.Arguments[i];

                codeCompare = string.CompareOrdinal(argx?.ToString(), argy?.ToString());
                if (codeCompare != 0)
                    return codeCompare;
            }

            return x.Arguments.Count.CompareTo(y.Arguments.Count);
```

All tests pass.

Do you think this will reduce allocations sufficiently?

**Comment by sharwell** [comment]
@YairHalberstadt That seems great but doesn't include source location. Is there a way to test source location too?

**Comment by YairHalberstadt** [comment]
What do you mean by testing source location?

**Comment by YairHalberstadt** [comment]
I believe that CanonicallyCompareDiagnostics would only run on diagnostics which have the same span, so given that the errorcode is the same, I think all that's important is the arguments.

**Comment by sharwell** [comment]
> I believe that CanonicallyCompareDiagnostics would only run on diagnostics which have the same span

I can't validate this, but the rest should be easy to submit for an initial PR. The compiler team can help review to determine if it's safe to omit the location from the comparison.

As for the code and the arguments, that seems like it would resolve the main points of this bug.

**Comment by YairHalberstadt** [comment]
https://github.com/dotnet/roslyn/pull/33930

**Comment by YairHalberstadt** [comment]
The fix has been merged. Should this issue be closed, or do you want to confirm the fix actually works?

**Comment by gafter** [comment]
Fixed in #33930.  Thanks, @YairHalberstadt
